packages/firestore/test/unit/specs/spec_test_runner.ts

```diff
@@ -1037,31 +1158,58 @@ export async function runSpec(
   config: SpecConfig,
   steps: SpecStep[]
 ): Promise<void> {
-  let runner: TestRunner;
-  if (usePersistence) {
-    runner = new IndexedDbTestRunner(name, config);
-  } else {
-    runner = new MemoryTestRunner(name, config);
+  console.log('Running spec: ' + name);
+  const platform = new TestPlatform(PlatformSupport.getPlatform());
+  let runners: TestRunner[] = [];
+  for (let i = 0; i < config.numClients; ++i) {
+    if (usePersistence) {
+      runners.push(new IndexedDbTestRunner(name, platform, config));
+    } else {
+      runners.push(new MemoryTestRunner(name, platform, config));
+    }
+    await runners[i].start();
   }
-  await runner.start();
+  let lastStep = null;
+  let count = 0;
   try {
-    await runner.run(steps);
+    await sequence(steps, async step => {
+      ++count;
+      lastStep = step;
+      return runners[step.clientIndex || 0].run(step);
+    });
+  } catch (err) {
+    console.warn(
+      `Spec test failed at step ${count}: ${JSON.stringify(lastStep)}`
+    );
+    throw err;
   } finally {
-    await runner.shutdown();
+    for (const runner of runners) {
+      await runner.shutdown();
+    }
+    if (usePersistence) {
+      await IndexedDbTestRunner.destroyPersistence();
+    } else {
+      await MemoryTestRunner.destroyPersistence();
+    }
   }
 }
 
 /** Specifies initial configuration information for the test. */
 export interface SpecConfig {
   /** A boolean to enable / disable GC. */
   useGarbageCollection: boolean;
+
+  /** The number of active clients for this test run. */
+  numClients: number;
 }
 
 /**
  * Union type for each step. The step consists of (mostly) exactly one `field`
  * set and optionally expected events in the `expect` field.
  */
 export interface SpecStep {
+  /** The index of the current client for multi-client spec tests. */
```

Comment:
Can you add "PORTING NOTE: Only used by web multi-tab tests." or similar to the new steps you added (except maybe shutdown).

Reply:
Added here and in a bunch of other places.